AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/DefaultKeyStorageInterface.dfy

```diff
@@ -585,78 +585,83 @@ module DefaultKeyStorageInterface {
                       identifier:=input.Identifier,
                       tableName:=ddbTableName));
 
         // SDKs/Smithy-Dafny/Custom Implementations of Storage MAY respond with None or an Empty Map.
         // .NET returns an empty map, Java returns None.
       :- Need(
-        ddbResponse.Responses.Some? && (2 == |ddbResponse.Responses.value|),
+        commitmentRes.Item.Some? && (0 < |commitmentRes.Item.value|),
+        Types.KeyStorageException(
+          message:=
+            "GetMutation's GetItem for the Mutation Commitment returned an empty result. "
+            + "Branch Key ID: " + input.Identifier
+            + "\tTable Name: " + ddbTableName));
+
+      // Get Mutation Index
+      var indexKey: DDB.Key := map[
+        Structure.BRANCH_KEY_IDENTIFIER_FIELD := DDB.AttributeValue.S(input.Identifier),
+        Structure.TYPE_FIELD := DDB.AttributeValue.S(Structure.MUTATION_INDEX_TYPE)
+      ];
+      var indexReq := DDB.GetItemInput(
+        Key := indexKey,
+        TableName := ddbTableName,
+        ConsistentRead :=  Some(true)
+      );
+      var indexRes? := ddbClient.GetItem(indexReq);
+      var indexRes :- indexRes?
+      .MapFailure((e: DDB.Error) =>
+                    wrapDdbException(
+                      e:=e,
+                      storageOperation:="GetMutation",
+                      ddbOperation:="GetItem",
+                      identifier:=input.Identifier,
+                      tableName:=ddbTableName));
+      // SDKs/Smithy-Dafny/Custom Implementations of Storage MAY respond with None or an Empty Map.
+      // .NET returns an empty map, Java returns None.
+      :- Need(
+        indexRes.Item.Some? && (0 < |indexRes.Item.value|),
         Types.KeyStorageException(
           message:=
-            "GetMutation: No items returned. "
+            "GetMutation's GetItem for the Mutation Index returned an empty result. "
             + "Branch Key ID: " + input.Identifier
             + "\tTable Name: " + ddbTableName));
 
       /** Process sensical DDB Response */
-      var lockCanidate := ddbResponse.Responses.value[0].Item;
+      ghost var lockCanidate := commitmentRes.Item;
       var lockItem: Option<Types.MutationCommitment> :-
-        MutationCommitmentFromOptionalItem(lockCanidate, input.Identifier, ddbTableName);
+        MutationCommitmentFromOptionalItem(commitmentRes.Item, input.Identifier, ddbTableName);
       assert lockItem.Some? ==> lockCanidate.Some? && Structure.MutationCommitmentAttribute?(lockCanidate.value);
 
-      var indexCanidate := ddbResponse.Responses.value[1].Item;
+      ghost var indexCanidate := indexRes.Item;
       var indexItem: Option<Types.MutationIndex> :-
-        MutationIndexFromOptionalItem(indexCanidate, input.Identifier, ddbTableName);
+        MutationIndexFromOptionalItem(indexRes.Item, input.Identifier, ddbTableName);
       assert indexItem.Some? ==> indexCanidate.Some? && Structure.MutationIndexAttribute?(indexCanidate.value);
```

Comment:
Same as above, I think that this is simpler as:

```dafny
      var indexItem: Types.MutationIndex :-
        StorageHelpers.MutationIndexFromItem(indexRes.Item.value, input.Identifier, ddbTableName);
      assert Structure.MutationIndexAttribute?(indexRes.Item.value);
```

Clearly you need to wrap it back up with `Some` for your output.
Since, I don't think you want to refactor that out for this as well.

Reply:
This method is used by Describe Mutation, which can return Nones if there is no Mutation in-flight.
That is why it is written this way; a none-result is valid.

Reply:
Also, Describe Mutation and Initialize Mutation will work if only the commitment exists, and assume the index was deleted to restart the mutation from the head of the table.